nomos-mix/message/src/layered_cipher.rs

```diff
@@ -0,0 +1,331 @@
+use std::marker::PhantomData;
+
+use rand_chacha::{
+    rand_core::{RngCore, SeedableRng},
+    ChaCha12Rng,
+};
+use sphinx_packet::{
+    constants::HEADER_INTEGRITY_MAC_SIZE,
+    crypto::STREAM_CIPHER_INIT_VECTOR,
+    header::{
+        keys::{HeaderIntegrityMacKey, StreamCipherKey},
+        mac::HeaderIntegrityMac,
+    },
+};
+
+use crate::{concat_bytes, parse_bytes};
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Invalid data length")]
+    InvalidDataLength,
+    #[error("Invalid params")]
+    InvalidParams,
+    #[error("Integrity MAC verification failed")]
+    IntegrityMacVerificationFailed,
+}
+
+type Result<T> = std::result::Result<T, Error>;
+
+/// A cipher to encrypt/decrypt a list of data of the same size using a list of keys.
+///
+/// The cipher performs the layered encryption.
+/// The following example shows the simplified output.
+/// - Input: [[data0, k0], [data1, k1]]
+/// - Output: encrypt(k0, [data0, encrypt(k1, [data1])])
+///
+/// The max number of layers is limited to the `max_layers` parameter.
+/// Even if the number of data and keys provided for encryption is smaller than `max_layers`,
+/// The cipher always produces the max-size output regardless of the number of data and keys provided,
+/// in order to ensure that all outputs generated by the cipher are the same size.
+///
+/// The cipher also provides the length-preserved decryption.
+/// Even if one layer of encryptions is decrypted, the length of decrypted data is
+/// the same as the length of the original data.
+/// For example:
+///   len(encrypt(k0, [data0, encrypt(k1, [data1])])) == len(encrypt(k1, [data1]))
+pub struct ConsistentLengthLayeredCipher<D: ConsistentLengthLayeredCipherData> {
```

Comment:
Sometimes there is no need to include the bound here, just in impl blocks. Maybe just `D` could be enough.

Reply:
Thanks [3d32942](https://github.com/logos-co/nomos-node/pull/897/commits/3d32942725a2d769e6a928b74426c97d287f11ab)